File: server_akka_scala/src/main/scala/info/rori/lunchbox/server/akka/scala/external/OcrClient.scala
package info.rori.lunchbox.server.akka.scala.external

import java.net.URL

import dispatch.Defaults._
import dispatch._
import play.api.libs.json._

import scala.concurrent.Future
import scala.util.matching.Regex

/**
  * Service für das Auslesen von Text aus einem Bild (OCR).
  */
object OcrClient extends HttpClient {

  implicit class RegexContext(sc: StringContext) {
    def r = new Regex(sc.parts.mkString, sc.parts.tail.map(_ => "x"): _*)
  }

  def doOCR(imageUrl: URL): Future[String] = {
    /*
      Tipp: Die HTTP-Requests an newocr.com lassen sich via Wireshark mitsamt aller Header loggen:
      tshark -V -Y "http.request and http.host contains newocr" > tshark.log
     */
    val body = Json.obj(
      "img_url" -> imageUrl.toString,
      "engine" -> "tesseract",
      "engine_args" -> Json.obj("lang" -> "deu"))

    val req = url("http://openocr:20080/ocr").POST
      .setContentType("application/json", "utf-8")
      .setBody(body.toString)

    Http(req OK (response => response.getResponseBody))
  }

}

File: backend-spring-kotlin/src/main/kotlin/lunchbox/util/ocr/OcrClient.kt
package lunchbox.util.ocr

import org.springframework.beans.factory.annotation.Value
import org.springframework.resilience.annotation.Retryable
import org.springframework.stereotype.Component
import org.springframework.web.client.RestClient
import org.springframework.web.client.body
import java.net.URL

/**
 * Führt Texterkennung (OCR) auf dem übergebenen Bild via OpenOCR aus.
 * <p>
 * OpenOCR wird in der Lunchbox via Docker bereitgestellt.
 */
@Component
class OcrClient(
  @Value("\${external.ocr.url:http://ocr:$OCR_PORT}")
  val ocrUrl: String,
) {
  @Retryable(maxRetries = 5, delay = 5000, multiplier = 2.0)
  fun doOCR(imageUrl: URL): String {
    val requestBody =
      mapOf(
        "img_url" to imageUrl,
        "engine" to "tesseract",
        "engine_args" to mapOf("lang" to "deu"),
      )

    return RestClient
      .create("$ocrUrl/url")
      .post()
      .body(requestBody)
      .retrieve()
      .body<String>() ?: ""
  }
}

const val OCR_PORT = 9292
